gremlin/ui/device.py

```diff
@@ -30,13 +31,108 @@
+from gremlin.types import InputType, PropertyType
+from gremlin.util import file_exists_and_is_accessible
+from gremlin.common import SingletonDecorator
+from gremlin.config import Configuration
 
 
 QML_IMPORT_NAME = "Gremlin.Device"
 QML_IMPORT_MAJOR_VERSION = 1
 
 
+@SingletonDecorator
+class DeviceDatabase:
+    """Provides button/axis to name mapping for known devices"""
+
+    def __init__(self) -> None:
+        self._load()
+
+    def _load(self) -> None:
+
+        db_file = "device_db.json"
+        if not file_exists_and_is_accessible(db_file):
+            return
+
+        load_successful = False
+        json_data = {}
+        with open(db_file) as hdl:
+            try:
+                decoder = json.JSONDecoder()
+                json_data = decoder.decode(hdl.read())
+                load_successful = True
+            except ValueError:
+                pass
+
+        if not load_successful:
+            logging.getLogger("system").error(
+                    f"There was an error loading {db_file}. "
+                    "Probably it's a syntax error, "
+                    "please fix the errors and restart the application.")
+
+        # create an empty device database if load was not successful
+        self.device_db = json_data if load_successful else {}
+
+    def input_name(self, input_name: str, input_map: Dict[str, str]) -> str:
+        """
+        Returns the input name based on the global configuration option
+        and availability of information about the input in
+        the device database.
+
+        If a device exists in the device database and it's mapping or
+        the input are not defined, return a default
+        input name, e.g. Axis 1 or Button 1.
+
+        Some devices have configurable number of buttons and/or axes so
+        adjusting device database information might be required in the future.
+        """
+        input_name_display_mode = Configuration().value(
+            "global", "input-names", "input-name-display-mode"
+            )
+
+        if input_name_display_mode == "None" or input_name not in input_map:
+            return input_name
+
+        db_input_name = input_map[input_name].strip()
+
+        # return default input name if input is defined, but empty
+        if len(db_input_name) == 0:
+            return input_name
+
+        # return configured input name style
+        if input_name_display_mode == "Mixed":
+            return f"{input_name} - {db_input_name}"
+        if input_name_display_mode == "Name":
+            return db_input_name
+
+    def get_mapping(self, device: dill.DeviceSummary) -> Dict[str, str]:
+        """Returns the button/axis <-> name mapping of a device"""
+        if device is None:
+            return None
+
+        for d in self.device_db["devices"]:
+            if (
+                d["product"] == device.product_id and
+                d["vendor"] == device.vendor_id
+            ):
+                if (
```

Comment:
Probably can be simplified to something like `d.get("mapping", None) in self.device_db["mapping|]` I would assume that `None` won't be found in the mapping dict.

Reply:
ok